onnxruntime/test/optimizer/graph_transform_test.cc

```diff
@@ -2350,5 +2353,184 @@ TEST_F(GraphTransformationTests, DynamicQuantizeMatMulTest) {
 
 #endif
 
+// LayerNormalization implementation is in contrib namespace (OnnxDomain 1), so
+// Without contib_ops enabled, we cannot parse the graph correctly.
+#ifndef DISABLE_CONTRIB_OPS
+
+// We used Opset 12 for testing to make sure we are not using GatherND OnnxDomain Opset 1.
+TEST_F(GraphTransformationTests, ComputationReductionTransformer_BasicCheck) {
+  auto model_uri = MODEL_FOLDER "computation_reduction_transformer.onnx";
+  std::shared_ptr<Model> model;
+  ASSERT_STATUS_OK(Model::Load(model_uri, model, nullptr, *logger_));
+  Graph& graph = model->MainGraph();
+  std::map<std::string, int> op_to_count = CountOpsInGraph(graph);
+
+  onnxruntime::GraphTransformerManager graph_transformation_mgr{1};
+  graph_transformation_mgr.Register(onnxruntime::make_unique<ComputationReductionTransformer>(), TransformerLevel::Level1);
+  ASSERT_STATUS_OK(graph_transformation_mgr.ApplyTransformers(graph, TransformerLevel::Level1, *logger_));
+
+  GraphViewer graph_viewer(graph);
+  const auto& node_topology_list = graph_viewer.GetNodesInTopologicalOrder();
+
+  Node* gathernd_node = nullptr;
+  for (auto node_index : node_topology_list) {
+    Node* p_node = graph.GetNode(node_index);
+    ASSERT_FALSE(p_node == nullptr);
+    if (p_node->OpType().compare("GatherND") == 0) {
+      gathernd_node = p_node;
+      const Node* layer_norm_node = graph.GetProducerNode(gathernd_node->MutableInputDefs()[0]->Name());
+      EXPECT_EQ(layer_norm_node->OpType(), "LayerNormalization");
+      EXPECT_EQ(layer_norm_node->Name(), "layer_norm_1");
+      const auto& consumers = graph.GetConsumerNodes(gathernd_node->MutableOutputDefs()[0]->Name());
+      EXPECT_EQ(consumers[0]->OpType(), "MatMul");
+      EXPECT_EQ(consumers[0]->Name(), "matmul_1");
+      break;
+    }
+  }
+
+  ASSERT_FALSE(gathernd_node == nullptr);
+}
+
+TEST_F(GraphTransformationTests, ComputationReductionTransformer_ResultCompare) {
+  auto model_uri = MODEL_FOLDER "computation_reduction_transformer.onnx";
+  std::shared_ptr<Model> model;
```

Comment:
I feel we should have a model for each supported case. It's not good to use a big model for all because it's super hard to debug when it fails.

Reply:
I agree that we should add some cases for each op swapping. While I think the bert-like structure test is still needed for correctness testing.